src/sinks/gcp/chronicle_unstructured.rs

```diff
@@ -0,0 +1,526 @@
+    };
+
+    Ok(Box::pin(healthcheck))
+}
+
+#[derive(Debug, Snafu)]
+pub enum ChronicleError {
+    #[snafu(display("Region or endpoint not defined"))]
+    RegionOrEndpoint,
+    #[snafu(display("You can only specify one of region or endpoint"))]
+    BothRegionAndEndpoint,
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "gcp_chronicle_unstructured")]
+impl SinkConfig for ChronicleUnstructuredConfig {
+    async fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
+        let creds = self.auth.build(Scope::MalachiteIngestion).await?;
+
+        let tls = TlsSettings::from_options(&self.tls)?;
+        let client = HttpClient::new(tls, cx.proxy())?;
+
+        let endpoint = self.create_endpoint("v2/unstructuredlogentries:batchCreate")?;
+
+        // For the healthcheck we see if we can fetch the list of available log types.
+        let healthcheck_endpoint = self.create_endpoint("v2/logtypes")?;
+
+        let healthcheck = build_healthcheck(client.clone(), &healthcheck_endpoint, creds.clone())?;
+        let sink = self.build_sink(client, endpoint, creds)?;
+
+        Ok((sink, healthcheck))
+    }
+
+    fn input(&self) -> Input {
+        Input::log()
+    }
+
+    fn sink_type(&self) -> &'static str {
+        NAME
+    }
+
+    fn acknowledgements(&self) -> Option<&AcknowledgementsConfig> {
+        Some(&self.acknowledgements)
+    }
+}
+
+impl ChronicleUnstructuredConfig {
+    fn build_sink(
+        &self,
+        client: HttpClient,
+        base_url: String,
+        creds: GcpAuthenticator,
+    ) -> crate::Result<VectorSink> {
+        use crate::sinks::util::service::ServiceBuilderExt;
+
+        let request = self.request.unwrap_with(&TowerRequestConfig {
+            rate_limit_num: Some(1000),
+            ..Default::default()
+        });
+
+        let batch_settings = self.batch.into_batcher_settings()?;
+
+        let partitioner = self.key_partitioner()?;
+
+        let svc = ServiceBuilder::new()
+            .settings(request, GcsRetryLogic)
+            .service(ChronicleService::new(client, base_url, creds));
+
+        let request_settings = RequestSettings::new(self)?;
+
+        let sink = GcsSink::new(svc, request_settings, partitioner, batch_settings);
+
+        Ok(VectorSink::from_event_streamsink(sink))
+    }
+
+    fn key_partitioner(&self) -> crate::Result<KeyPartitioner> {
+        Ok(KeyPartitioner::new(self.log_type.clone()))
+    }
+
+    fn create_endpoint(&self, path: &str) -> Result<String, ChronicleError> {
+        Ok(format!(
+            "{}/{}",
+            match (&self.endpoint, self.region) {
+                (Some(endpoint), None) => endpoint.trim_end_matches('/'),
+                (None, Some(region)) => region.endpoint(),
+                (Some(_), Some(_)) => return Err(ChronicleError::BothRegionAndEndpoint),
+                (None, None) => return Err(ChronicleError::RegionOrEndpoint),
```

Comment:
It seems like this logic is configuration based , would it be beneficial to add a unit test case for it ?

Reply:
Not sure I understand what you mean?